src/libpgmoneta/message.c

```diff
@@ -552,11 +562,518 @@ pgmoneta_create_startup_message(char* username, char* database, struct message**
+/*       { */
+/*          return pgmoneta_get_column(i, tuple); */
+/*       } */
+/*    } */
+
+/*    return NULL; */
+/* } */
+
+int
+pgmoneta_query_execute(int socket, struct message* msg, struct query** query)
+{
+   int status;
+   int fd = -1;
+   bool cont;
+   int cols;
+   char* name = NULL;
+   struct message* tmsg = NULL;
+   char* content = NULL;
+   struct message* reply = NULL;
+   struct query* q = NULL;
+   struct tuple* current = NULL;
+   size_t data_size;
+   void* data = pgmoneta_memory_dynamic_create(&data_size);
+   size_t offset = 0;
+
+   *query = NULL;
+
+   status = pgmoneta_write_message(NULL, socket, msg);
+   if (status != MESSAGE_STATUS_OK)
+   {
+      goto error;
+   }
+
+   cont = true;
+   while (cont)
+   {
+      status = pgmoneta_read_block_message(NULL, socket, &reply);
+
+      if (status == MESSAGE_STATUS_OK)
+      {
+         data = pgmoneta_memory_dynamic_append(data, data_size, reply->data, reply->length, &data_size);
+
+         if (pgmoneta_has_message('Z', data, data_size))
+         {
+            cont = false;
+         }
+      }
+      else if (status == MESSAGE_STATUS_ZERO)
+      {
+         SLEEP(1000000L);
+      }
+      else
+      {
+         goto error;
+      }
+
+      pgmoneta_free_message(reply);
+      reply = NULL;
+   }
+
+   if (pgmoneta_has_message('E', data, data_size))
+   {
+      goto error;
+   }
+
+   if (pgmoneta_extract_message_from_data('T', data, data_size, &tmsg))
+   {
+      goto error;
+   }
+
+   cols = get_number_of_columns(tmsg);
+
+   q = (struct query*)malloc(sizeof(struct query));
+   memset(q, 0, sizeof(struct query));
+
+   q->number_of_columns = cols;
+
+   for (int i = 0; i < cols; i++)
+   {
+      if (get_column_name(tmsg, i, &name))
+      {
+         goto error;
+      }
+
+      memcpy(&q->names[i][0], name, strlen(name));
+
+      free(name);
+      name = NULL;
+   }
+
+   while (offset < data_size)
+   {
+      offset = pgmoneta_extract_message_offset(offset, data, &msg);
+
+      if (msg != NULL && msg->kind == 'D')
+      {
+         struct tuple* dtuple = NULL;
+
+         create_D_tuple(cols, msg, &dtuple);
+
+         if (q->tuples == NULL)
+         {
+            q->tuples = dtuple;
+         }
+         else
+         {
+            current->next = dtuple;
+         }
+
+         current = dtuple;
+      }
+
+      pgmoneta_free_copy_message(msg);
+      msg = NULL;
+   }
+
+   *query = q;
+
+   pgmoneta_free_copy_message(tmsg);
+   pgmoneta_memory_dynamic_destroy(data);
+
+   free(content);
+
+   return 0;
+
+error:
+
+   pgmoneta_disconnect(fd);
+
+   pgmoneta_free_message(msg);
+   pgmoneta_free_copy_message(tmsg);
+   pgmoneta_memory_dynamic_destroy(data);
+
+   free(content);
+
+   return 1;
+}
+
+bool
+pgmoneta_has_message(char type, void* data, size_t data_size)
+{
+   int offset;
+
+   offset = 0;
+
+   while (offset < data_size)
+   {
+      char t = (char)pgmoneta_read_byte(data + offset);
```

Comment:
I'm not familiar with the mechanism here. Is there some agreement on where the type of message will be within the message?

Reply:
Protocol again